eos-autoupdater/main.c

```diff
@@ -274,6 +316,153 @@ update_stamp_file (guint64 last_successful_update_secs,
   g_debug ("Wrote stamp file. Next update at %s", next_update);
 }
 
+/**
+ * read_poll_results_file:
+ *
+ * Reads the stored poll results file and returns a new #PollResults
+ * instance.
+ *
+ * Returns: (transfer full) (nullable): a newly allocated #PollResults
+ *  or %NULL if the results file could not be read
+ */
+static PollResults *
+read_poll_results_file (void)
+{
+  const gchar *state_dir = get_state_dir ();
+  g_autofree gchar *results_path = NULL;
+  g_autoptr(GFile) results_file = NULL;
+  g_autoptr(GBytes) results_contents = NULL;
+  g_autoptr(GVariant) results_variant = NULL;
+  g_autoptr(GError) error = NULL;
+  guint64 last_changed_usecs = 0;
+  const gchar *update_refspec = NULL;
+  const gchar *update_id = NULL;
+
+  results_path = g_build_filename (state_dir, POLL_RESULTS_NAME, NULL);
+  results_file = g_file_new_for_path (results_path);
+  results_contents = g_file_load_bytes (results_file, NULL, NULL, &error);
+  if (results_contents == NULL)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND)) {
+        critical (EOS_UPDATER_STAMP_ERROR_MSGID,
+                  "Failed to read autoupdater poll results file: %s",
+                  error->message);
+      }
+
+      return NULL;
+    }
+
+  results_variant = g_variant_new_from_bytes (G_VARIANT_TYPE ("a{sv}"),
+                                              results_contents,
+                                              FALSE);
+  g_variant_lookup (results_variant, "LastChangedUsecs", "t", &last_changed_usecs);
+  g_variant_lookup (results_variant, "UpdateRefspec", "&s", &update_refspec);
+  g_variant_lookup (results_variant, "UpdateID", "&s", &update_id);
```

Comment:
Any/All of these lookups may fail (and return `FALSE`) if the file is corrupt. Is that OK, or do you want to return `NULL` from the function? In any case, perhaps a comment about how invalid files are handled would clarify things.

Reply:
Yes, a further comment would help. Originally I had all of these catching and handling the `NULL` results, but then I added the `poll_results_new` helper so they could all be handled in one place.

I do believe this should try to handle a corrupted data file. Even though it might hide problems in the code, I think it would be better than having the autoupdater get wedged. I'll add some warnings here so maybe someone will notice them in the logs if this code has persistent issues.

Reply:
Sorry, now I see what you're saying. The helper is already returning `NULL` if the file doesn't exist, so should it do the same if there are fields missing? I'm worried about the autoupdater never applying a user visible update if there are issues in serializing or deserializing the state. However, I don't know if it will do the right thing with partial results, so I think maybe it is best to return `NULL` with a warning when that happens.

Reply:
My comment was (not worded very well and) about whether it’s OK for the function to ignore partial errors (some but not all expected fields missing/incorrectly typed), or whether the function should return `NULL` on any error and only return success if every field is found and is valid.

The question of whether the update should be applied if this function returns an error is a separate one. I agree that we should avoid a corrupt file here blocking an update, so printing a warning and carrying on anyway seems appropriate.